src/pydvl/influence/frameworks/torch_differentiable.py

```diff
@@ -33,93 +26,164 @@ def flatten_gradient(grad):
     return torch.cat([el.reshape(-1) for el in grad])
 
 
-class TorchTwiceDifferentiable(TwiceDifferentiable):
+def solve_linear(matrix: torch.Tensor, b: torch.Tensor):
+    """Computes the solution of a square system of linear equations"""
+    return torch.linalg.solve(matrix, b)
+
+
+def as_tensor(a: Any, ensure=True, **kwargs):
+    """Converts an array into a torch tensor"""
+    if ensure and not isinstance(a, torch.Tensor):
+        logger.warning("Converting tensor to type torch.Tensor.")
+    return torch.as_tensor(a, **kwargs)
+
+
+def stack_tensors(a: Sequence[torch.Tensor], **kwargs):
+    """Stacks a sequence of tensors into a single torch tensor"""
+    return torch.stack(a, **kwargs)
+
+
+def einsum(equation, *operands):
+    """Sums the product of the elements of the input :attr:`operands` along dimensions specified using a notation
+    based on the Einstein summation convention.
+    """
+    return torch.einsum(equation, *operands)
+
+
+def identity_tensor(dim: int):
+    return torch.eye(dim, dim)
+
+
+def mvp(
+    grad_xy: torch.Tensor,
+    v: torch.Tensor,
+    backprop_on: torch.Tensor,
+    progress: bool = False,
+) -> torch.Tensor:
+    """
+    Calculates second order derivative of the model along directions v.
+    This second order derivative can be selected through the backprop_on argument.
+
+    :param grad_xy: an array [P] holding the gradients of the model
+        parameters wrt input $x$ and labels $y$, where P is the number of
+        parameters of the model. It is typically obtained through
+        self.grad.
+    :param v: An array ([DxP] or even one dimensional [D]) which
+        multiplies the matrix, where D is the number of directions.
+    :param progress: True, iff progress shall be printed.
+    :param backprop_on: tensor used in the second backpropagation (the first
+        one is along $x$ and $y$ as defined via grad_xy).
+    :returns: A matrix representing the implicit matrix vector product
+        of the model along the given directions. Output shape is [DxP] if
+        backprop_on is None, otherwise [DxM], with M the number of elements
+        of backprop_on.
+    """
+    v = as_tensor(v, ensure=False)
+    if v.ndim == 1:
+        v = v.unsqueeze(0)
+
+    z = (grad_xy * Variable(v)).sum(dim=1)
+
+    mvp = []
+    for i in maybe_progress(range(len(z)), progress, desc="MVP"):
+        mvp.append(
+            flatten_gradient(autograd.grad(z[i], backprop_on, retain_graph=True))
+        )
+    mvp = torch.stack([grad.contiguous().view(-1) for grad in mvp])
+    return mvp.detach()  # type: ignore
+
+
+class TorchTwiceDifferentiable(TwiceDifferentiable[torch.Tensor, nn.Module]):
     """
-    Calculates second-derivative matrix vector products (Mvp) of a pytorch torch.nn.Module
+    Calculates second-derivative of a model wrt. a given loss
```

Comment:
This generic description belongs in the Base class, I'd say. Also, I would write something like
"Wraps a differentiable model and loss and provides methods to compute the second derivative of the loss wrt. the model parameters" (if this is what is happening. At any rate the derivative of the model wrt a loss doesn't really make sense I think)